catalog/controller/payment/securesubmit.php

```diff
@@ -40,80 +144,305 @@ public function index() {
+
+        // get these values from the DB
+        $securesubmit_fraud_enable          = $this->config->get('securesubmit_fraud_enable');
+        $securesubmit_fraud_message         = $this->config->get('securesubmit_fraud_message');
+        $securesubmit_fraud_fail            = $this->config->get('securesubmit_fraud_fail');
+        $securesubmit_fraud_time            = $this->config->get('securesubmit_fraud_time');
+
+        // if invalid or null set defaults for each
+        if ($securesubmit_fraud_enable === null){ // set default
+            $securesubmit_fraud_enable      = self::SECURESUBMIT_FRAUD_ENABLE_DEFAULT;
+        }
+        if ($securesubmit_fraud_message === null){ // set default
+            $securesubmit_fraud_message     = self::SECURESUBMIT_FRAUD_MESSAGE_DEFAULT;
+        }
+        if ((int)$securesubmit_fraud_fail < 1){ // set default
+            $securesubmit_fraud_fail        = self::SECURESUBMIT_FRAUD_FAIL_DEFAULT;
+        }
+        if ((int)$securesubmit_fraud_time < 1){ // set default
+            $securesubmit_fraud_time        = self::SECURESUBMIT_FRAUD_TIME_DEFAULT;
+        }
+
+        // validation done lets set our properties
+        $this->securesubmit_fraud_enable    = (bool)$securesubmit_fraud_enable === 'on';
+        $this->securesubmit_fraud_message   = filter_var(trim($securesubmit_fraud_message), FILTER_SANITIZE_STRING);
+        $this->securesubmit_fraud_fail      = (int)$securesubmit_fraud_fail;
+        $this->securesubmit_fraud_time      = (int)$securesubmit_fraud_time*self::MINUTE_IN_SECONDS;
+
+        // set a session var based on the users IP and if not set instantiate the count to 0
+        $this->securesubmit_fraud_fail_var  = "HeartlandHPS_FailCount_" . $this->getIP();
+        $this->sessionVar('count');
+    }
+
+    /**
+     * @return string
+     */
+    private function getIP(){
+
+        if(array_key_exists("HTTP_X_FORWARDED_FOR",$_SERVER) && $_SERVER["HTTP_X_FORWARDED_FOR"] != ""){
+            $IParray = array_values(array_filter(explode(',',$_SERVER['HTTP_X_FORWARDED_FOR'])));
+            $IP = end($IParray);
+        }else{
+            $IP = $_SERVER["REMOTE_ADDR"];
+        }
+        return md5($IP);
+    }
+
+    /**
+     * Open cart sets session variables to never timeout while the browser is open
+     * Heartland needs these values to die at a time defined in settings
+     * this function works by checking the last time a var was set and  resetting if they have timed out
+     * @param null|string   $key
+     * @param null|string   $val
+     *
+     * @return mixed
+     */
+    private function sessionVar($key = null, $val = null){
+        /*
+         * I need to set a timebomb on the following variables
+         * $this->session->data[$this->securesubmit_fraud_fail_var]['count']
+         * $this->session->data[$this->securesubmit_fraud_fail_var]['previous']
+         * $this->session->data[$this->securesubmit_fraud_fail_var]['lastErrorTime']
+         */
```

Comment:
@ttnnkkrr Just to confirm, is this session data unique to each user session? It doesn't act like a cache, right?